### PR TITLE
Okx: Fix panic during shutdown due to race

### PR DESCRIPTION
**Pull request by gbjk**
Error:
```
panic: runtime error: slice bounds out of range [:13340] with capacity 8192
bufio.(*Reader).Read(0x14000038540, {0x1400059cffc?, 0x4?, 0x0?})
        /usr/local/go/src/bufio/bufio.go:250 +0x33c
github.com/gorilla/websocket.(*messageReader).Read(0x1400098bb78, {0x1400059cffc, 0x4, 0x4})
        /Users/gbjk/go/pkg/mod/github.com/gorilla/websocket@v1.5.0/conn.go:1050 +0x208
io.ReadAll({0x12f36c618, 0x1400098bb78})
        /usr/local/go/src/io/io.go:701 +0xe4
io/ioutil.ReadAll(...)
        /usr/local/go/src/io/ioutil/ioutil.go:27
github.com/gorilla/websocket.(*Conn).ReadMessage(0x3?)
        /Users/gbjk/go/pkg/mod/github.com/gorilla/websocket@v1.5.0/conn.go:1097 +0x54
github.com/thrasher-corp/gocryptotrader/exchanges/stream.(*WebsocketConnection).ReadMessage(0x140006622d0)
        /Users/gbjk/go/pkg/mod/github.com/gbjk/gocryptotrader@v0.0.0-20230619070715-ae6f283f6be6/exchanges/stream/websocket_connection.go:217 +0x30
github.com/thrasher-corp/gocryptotrader/exchanges/okx.(*Okx).wsFunnelConnectionData(0x140019321e0?, {0x106909450, 0x140006622d0})
        /Users/gbjk/go/pkg/mod/github.com/gbjk/gocryptotrader@v0.0.0-20230619070715-ae6f283f6be6/exchanges/okx/okx_websocket.go:346 +0x94
created by github.com/thrasher-corp/gocryptotrader/exchanges/okx.(*Okx).WsConnect
        /Users/gbjk/go/pkg/mod/github.com/gbjk/gocryptotrader@v0.0.0-20230619070715-ae6f283f6be6/exchanges/okx/okx_websocket.go:234 +0x134
exit status 2
```
This happens when there's a race in calls to bufio because it over-reads. See [this comment](https://github.com/golang/go/issues/42289#issuecomment-723393783) Detected using go -race:
```
WARNING: DATA RACE
Read at 0x00c000818bc0 by goroutine 2156:
  github.com/gorilla/websocket.(*Conn).NextReader()
      /Users/gbjk/go/pkg/mod/github.com/gorilla/websocket@v1.5.0/conn.go:1000 +0x38
  github.com/gorilla/websocket.(*Conn).ReadMessage()
      /Users/gbjk/go/pkg/mod/github.com/gorilla/websocket@v1.5.0/conn.go:1093 +0x28
  github.com/thrasher-corp/gocryptotrader/exchanges/stream.(*WebsocketConnection).ReadMessage()
      /Users/gbjk/go/pkg/mod/github.com/gbjk/gocryptotrader@v0.0.0-20230619070715-ae6f283f6be6/exchanges/stream/websocket_connection.go:217 +0x44
  github.com/thrasher-corp/gocryptotrader/exchanges/okx.(*Okx).wsFunnelConnectionData()
```

Because we started a new wsFunnelConnectionData for each re-connect.

This bug might apply to other exchanges.

## Type of change

Please delete options that are not relevant and add an `x` in `[]` as item is complete.

- [x] Bug fix (non-breaking change which fixes an issue)

## How has this been tested

- [x] go test ./... -race
- [x] golangci-lint run

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation and regenerated documentation via the documentation tool
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally and on Github Actions/AppVeyor with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by gbjk** [comment]
This might have introduced an issue with waitgroups during shutdown.
Investigating now.

**Comment by gbjk** [comment]
😂  Okay, yeah, that was obvious in hindsight.
Fixed now.

**Comment by gbjk** [comment]
I'm still seeing waitgroup issues locally. I will get to the bottom of them and report back.

**Comment by gbjk** [comment]
Okay, I think this last commit makes a lot more sense.
I can't see any reason to have a goro feeding the multiplexer. Might as well just go straight to the channel.

Tested with Normal shutdowns and with forced disconnects.

Also sets precedent for how an exchange should wrap the Base.Shutdown, something which probably should be needed more often!

File: exchanges/okx/okx_websocket.go
Comment (gloriousCode) [reply exchanges/okx/okx_websocket.go:350]: Now to go further, I wrote the following:
```go
var indexPerMessage, indexPerGoRoutine int64

// wsReadData sends msgs from public and auth websockets to data handler
func (ok *Okx) wsReadData(ws stream.Connection) {
	defer ok.Websocket.Wg.Done()
	for {
		resp := ws.ReadMessage()
		if resp.Raw == nil {
			return
		}
		indexPerMessage++
		localIndexPerMessage := indexPerMessage
		// Ensure that websocket shutdown will block until WsHandleData returns, so that the multiplexer is safe from new blocking channel sends.
		ok.Websocket.Wg.Add(1)
		go func(msg []byte, passedInIndex int64) {
			indexPerGoRoutine++
			defer ok.Websocket.Wg.Done()
			if err := ok.WsHandleData(msg); err != nil {
				ok.Websocket.DataHandler <- err
			}
			if passedInIndex != indexPerGoRoutine {
				log.Warnf(log.ExchangeSys, "out of order index indexPerGoRoutine %v passedInIndex %v", indexPerGoRoutine, passedInIndex)
			}
		}(resp.Raw, localIndexPerMessage)
	}
}

```
Maybe I've written a bad test. This is without an authconnection. What I'm doing here is that `ReadMessage` is sequential, so the `indexPerMessage` denotes the latest update. I localise it to prevent anything weird, then pass it by value to the goroutine. In theory, these should always match if its FIFO right? But they don't when there is sufficient load

Further, I also put a `panic` when an orderbook checksum failed validation and it was able to be hit only when there was the goroutine. So we can't process updates out of order

**Review by gloriousCode — CHANGES_REQUESTED** [review]
Nice work on the investigation and fixes!

File: exchanges/okx/okx_wrapper.go
Comment (gbjk) [reply exchanges/okx/okx_wrapper.go:234]: Sometimes I avoid tailing a err return, because it's not the intrinsic return value of a func.
It's no more meaningful than the previous SetupNewConnection.
This way it's clear for future maintainers that if they want to add something that happens "at the end" or "close to the end" they don't put it above this SetupNewConn, or feel scared to turn `return funcCall` back into an if statement.

I'm not committed to it until death, but it's deliberate and reasoned. 😄 

Have I swayed you?

File: exchanges/okx/okx_websocket.go
Comment (gbjk) [reply exchanges/okx/okx_websocket.go:350]: Okay. I'll institute a FIFO.
I would never be worried about blowing up the number of goros.
I've frequently run with hundreds of thousands with no issue at all. It's literally one of the cornerstones of golang.
In this case they're not even long running, so I wouldn't hesitate to do this at all for that reason.

The reason I don't want to FIFO everything is because we don't want to block processing incoming order txs because of resource contention for orderbook and ticker.
Like you say, okx can be busy. That's the exact reason I don't want to put everything in the same fifo.

I think I'll look for a compromise downstream.

Note: Pretty sure we can't/shouldn't wg.Add before reading, because it's a blocking call, right?

And one final note: I get okx checksum errors constantly right now:
```
2023-06-30T12:26:43.169+0700    ERROR   Websocket error {"stategy": "xmm", "exchange": "Okx", "error": "Okx BTC-USDT spot orderbook data integrity compromised Reason: [Okx order book update checksum failed for pair BTC-USDT]"}
2023-06-30T12:26:44.311+0700    ERROR   Websocket error {"stategy": "xmm", "exchange": "Okx", "error": "Okx BTC-USDT spot orderbook data integrity compromised Reason: [Okx order book update checksum failed for pair BTC-USDT]"}
2023-06-30T12:26:51.119+0700    ERROR   Websocket error {"stategy": "xmm", "exchange": "Okx", "error": "Okx BTC-USDT spot orderbook data integrity compromised Reason: [Okx order book update checksum failed for pair BTC-USDT]"}
2023-06-30T12:26:57.639+0700    ERROR   Websocket error {"stategy": "xmm", "exchange": "Okx", "error": "Okx BTC-USDT spot orderbook data integrity compromised Reason: [Okx order book update checksum failed for pair BTC-USDT]"}
```
It's on my list to fix. They don't usually start straight away but get worse after a few hours.
Maybe I'll role that into this fix if at all possible. I suspect it's still related to dropped packets, despite a very good connection here.

Comment (gbjk) [reply exchanges/okx/okx_websocket.go]: I've confirmed you contention and issue.

I've switched to blocking processing for this PR.

I'll work on breaking out the messages which take longer another day.
It's mostly orderbook which can block for up to 4ms. We definitely can't have that.

**Review by gloriousCode — COMMENTED** [review]
Thanks for making those changes. I've tested this thoroughly and think its a good fix. I did come across an issue when it comes to CLI command `disableexchange` which I'll look in to as it applies to all exchanges.

Once you merge `master` branch, this should be good to go

**Comment by gbjk** [comment]
Rebased on master. I despise merging upstreams back in because of the disconnect in conflict resolution and the magical merge commit including changes from neither source intermingled.
Anyway, ready to go.

FYI: I didn't get notification of the merge label being added, so it's good you replied to it today 👍

File: exchanges/okx/okx_wrapper.go
Comment (gbjk) [reply exchanges/okx/okx_wrapper.go:229]: Okay, I can change this then. Will do so now.

**Review by gloriousCode — APPROVED** [review]
tACK! Thank you for this contribution!

**Review by shazbert — CHANGES_REQUESTED** [review]
Thanks for checking this out and opening this PR. Just some things I noticed on standard testing.

File: exchanges/okx/okx_websocket.go
Comment (gbjk) [reply exchanges/okx/okx_websocket.go:346]: Yeah, you're right. Hangover from earlier when this was a goro.

File: exchanges/okx/okx_websocket.go
Comment (gbjk) [reply exchanges/okx/okx_websocket.go:1685]: I don't agree that's safe to do, because Run() is likely to read from the NEW m.shutdown.

However we could remedy that by putting a m.shutdown = make(chan bool) in Run, right?

Comment (gbjk) [reply exchanges/okx/okx_websocket.go:1685]: Option 1: Localise a var to the shutdown chan in Run before run starts. Still has risk of race though.
Option 2: Have Run re-make the chan itself after consuming the close. Only limitation: Single consumer pattern. Anything else watching shutdown would get scuppered. I think that's safe given size and scope of wsRequestDataChannelsMultiplexer

So I'll fix with Option 2

File: exchanges/okx/okx_wrapper.go
Comment (gbjk) [reply exchanges/okx/okx_wrapper.go:217]: Having it in the Connect func was the whole problem before.
The multiplexer should survive disconnects.

Why would we want to run the multiplexer again?

**Review by shazbert — APPROVED** [review]
tACK. Thanks for the changes.

**Review by thrasher- — APPROVED** [review]
LGTM! 🏎️ 🔧

**Comment by Vankeee** [comment]
@gloriousCode @gbjk Hi guys, I find this data race is still occur when I re-Subscribe okx without shutdown the connection.
I use `go okxExch.Websocket.Connect()` to manange the ws.

**Comment by gbjk** [comment]
@Vankeee Ack and Thanks for reporting 🙏 
I've got some more work in the pipeline affecting subs throughout; It's possible I've addressed this already in that.
I'll create a behaviour test for this soon and get you a fix.

**Comment by Vankeee** [comment]
> @Vankeee Ack and Thanks for reporting 🙏 I've got some more work in the pipeline affecting subs throughout; It's possible I've addressed this already in that. I'll create a behaviour test for this soon and get you a fix.

Thanks!

**Comment by gbjk** [comment]
@Vankeee 
Can you clarify a couple of things for me:
* What commit of GCT master are you using?
* You said that when you resubscribe without shut down the connection, and that you use 
```
go okxExch.Websocket.Connect()
```
  * Is there a reason you're connecting in a go routine?
  * Are you checking the error return value from calling Connect?
* You've said that you get a panic. Could you please share the panic and stack trace for me?

My initial issue is that calling Connect twice should have returned you already Connected.
If that *didn't* happen, and the connection was disconnected, and you called it, you should have gotten something more like:
`subscription failure, channel already subscribed for {Key:{Channel:trades Currency:BTC-USD-221007 Asset:futures} Channel:trades Currency:BTC-USD-221007 Asset:futures Params:map[] State:0}`

If I spin up a load of Connects I can get a race, but I cannot get a panic.

Thanks for your help

**Comment by gbjk** [comment]
The race I was getting was due to a testing artifact: `WsConnect` ( which shouldn't be called directly ) is currently used internally in the _tests_ setupWS and bypasses the mutex protection.

TL;DR: Not sure how to get this to race, yet.

**Comment by gbjk** [comment]
I don't have any way forwards on this right now.
* The version of okx_websocket.go being used wasn't after the fixes to the race
* okx_websocket.go had been modified
* We haven't managed to reproduce this elsewhere

Next steps should be to reproduce with the latest version of GCT.

**Pull request merged**
